core/Enclave/enclave_hsm.cpp

```diff
@@ -147,653 +157,2072 @@ static uint32_t sgx_get_gcm_ciphertext_size(const sgx_aes_gcm_data_ex_t *gcm_dat
-        ret = SGX_SUCCESS;
-        goto out;
+    ret = sm4_decrypt(dec_key,
+                      cipherblob,
+                      plaintext_len,
+                      plaintext,
+                      iv,
+                      SGX_SM4_IV_SIZE,
+                      aad,
+                      aad_len,
+                      keyspec);
+    if (SGX_SUCCESS != ret) {
+        printf("error decrypting encrypted text\n");
     }
 
-    if (plaintext == NULL || plaintext_len < real_plaintext_len) {
-        printf("ecall rsa_decrypt plaintext_len(%d) is smaller than expected: %d.\n", plaintext_len, real_plaintext_len);
-        ret = SGX_ERROR_INVALID_PARAMETER;
-        goto out;
+    memset_s(dec_key, sizeof(dec_key), 0, sizeof(dec_key));
+    free(dec_key);
+
+    return ret;
+}
+
+/**
+ * @brief generate aes_gcm key with openssl api
+ * @param cmk_blob storage key information
+ * @param cmk_blob_size the size of cmk_blob
+ * @param req_blob_size the gcm data size
+ * @param key_size the size of aes key
+ * @return sgx_status_t
+ */
+sgx_status_t enclave_create_aes_key(uint8_t *cmk_blob, uint32_t cmk_blob_size, uint32_t *req_blob_size, uint32_t key_size)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+    if(key_size == UINT32_MAX) {
+        return SGX_ERROR_UNEXPECTED;
     }
+    uint32_t real_blob_len = ehsm_calc_gcm_data_size(0, key_size);
 
-    ret = sgx_rsa_priv_decrypt_sha256(pri_key, plaintext, &real_plaintext_len, ciphertext, ciphertext_len);
+    if (real_blob_len == UINT32_MAX) {
+        return SGX_ERROR_UNEXPECTED;
+    }
+    if (req_blob_size != NULL) {
+        *req_blob_size = real_blob_len;
+        return SGX_SUCCESS;
+    }
+    if (cmk_blob == NULL || cmk_blob_size != real_blob_len) {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+    uint8_t* tmp = (uint8_t *)malloc(key_size);
+    if (tmp == NULL) {
+        return SGX_ERROR_OUT_OF_MEMORY;
+    }
+    ret = sgx_read_rand(tmp, key_size);
     if (ret != SGX_SUCCESS) {
-        printf("ecall rsa_decrypt sgx_rsa_priv_decrypt_sha256 failed: %d.\n", ret);
-        goto out;
+        free(tmp);
+        return ret;
     }
+    ret = ehsm_gcm_encrypt(&g_domain_key, key_size, tmp, 0, NULL, cmk_blob_size, (sgx_aes_gcm_data_ex_t *)cmk_blob);
 
-out:
-    if (pri_key)
-        sgx_free_rsa_key(pri_key, SGX_RSA_PRIVATE_KEY, RSA_OAEP_3072_MOD_SIZE, RSA_OAEP_3072_EXP_SIZE);
-    memset_s(&rsa_key, sizeof(rsa_params_t), 0, sizeof(rsa_params_t));
+    memset_s(tmp, key_size, 0, key_size);
 
+    free(tmp);
     return ret;
 }
 
+/**
+ * @brief keybloblen = 0: get keybloblen; keybloblen != 0: create rsa key
+ * public key storages with plaintext and private key will be encrypted with domain_key
+ * @param cmk_blob
+ * @param cmk_blob_size
+ * @param req_blob_size
+ * @param keyspec
+ * @param padding_mode
+ * @return sgx_status_t
+ */
+sgx_status_t enclave_create_rsa_key(uint8_t *cmk_blob,
+                                    uint32_t cmk_blob_size,
```

Comment:
suggest to combine the symmetric createkey as: enclave_create_key

asymmetric related create key API to: enclave_create_keypair